src/resmom/linux/alps.c

```diff
@@ -6397,129 +6665,916 @@ alps_inventory(void)
+						   BASIL_VAL_VERSION_1_7) != NULL) {
+						basil_1_7_supported = 1;
+						snprintf(basilversion_system, sizeof(basilversion_system),
+							 BASIL_VAL_VERSION_1_7);
+					} else
+						basil_1_7_supported = 0;
+				}
+			} else {
+				/* Wrong method in the response. */
+				sprintf(log_buffer, "Wrong method, expected: %d but "
+					"got: %d", basil_method_query, brp_eng->method);
+				log_event(PBSEVENT_DEBUG, PBS_EVENTCLASS_NODE,
+					LOG_DEBUG, __func__, log_buffer);
+			}
+		} else {
+			/* There was an error in the BASIL response. */
+			sprintf(log_buffer, "Error in BASIL response: %s", brp_eng->error);
+			log_event(PBSEVENT_DEBUG, PBS_EVENTCLASS_NODE, LOG_DEBUG,
+				__func__, log_buffer);
+		}
+	}
+
+	free_basil_response_data(brp_eng);
+}
+
+/**
+ * @brief Process the System (BASIL 1.7) Query Response. This includes creation
+ *	  of KNL vnodes.
+ *
+ * @return void
+ */
+void
+system_to_vnodes_KNL(void)
+{
+	basil_response_query_system_t *sys_knl;
+
+	if (!basil_1_7_supported)
+		return;
+
+	/* System 1.7 Query failed to get KNL Node information from ALPS. */
+	if (!brp_knl)
+		return;
+
+	if (brp_knl->method != basil_method_query) {
+		snprintf(log_buffer, sizeof(log_buffer), "Wrong method: %d",
+			brp_knl->method);
+		log_event(PBSEVENT_DEBUG, PBS_EVENTCLASS_NODE, LOG_DEBUG, __func__,
+			log_buffer);
+		return;
+	}
+
+	if (brp_knl->data.query.type != basil_query_system) {
+		snprintf(log_buffer, sizeof(log_buffer), "Wrong query type: %d",
+			brp_knl->data.query.type);
+		log_event(PBSEVENT_DEBUG, PBS_EVENTCLASS_NODE, LOG_DEBUG, __func__,
+			log_buffer);
+		return;
+	}
+
+	if (*brp_knl->error != '\0') {
+		snprintf(log_buffer, sizeof(log_buffer), "Error in BASIL response: %s",
+			brp_knl->error);
+		log_event(PBSEVENT_DEBUG, PBS_EVENTCLASS_NODE, LOG_DEBUG, __func__,
+			log_buffer);
+		return;
+	}
+
+	sys_knl = &brp_knl->data.query.data.system;
+
+	create_vnodes_KNL(sys_knl);
+
+	free_basil_response_data(brp_knl);
+}
+
+/**
+ *
+ * @brief Create KNL vnodes.
+ * 	  'vnlp' is a global pointer that gets freed in process_hup() (mom_main.c)
+ * 	  during a MoM restart. The vnode state gets cleared and we repeat the vnode
+ *	  creation cycle i.e. creation of non-KNL vnodes (in inventory_to_vnodes())
+ *	  followed by KNL vnodes in this function.
+ *
+ * @param[in] sys_knl ALPS BASIL System Query response.
+ *
+ * @return void
+ */
+static void
+create_vnodes_KNL(basil_response_query_system_t *sys_knl)
+{
+	char	*attr, *arch;
+	char	vname[VNODE_NAME_LEN];
+	char	utilBuffer_knl[(UTIL_BUFFER_LEN * sizeof(char))];
+	int	ncpus_per_knl;
+	int	node_idx = 0;
+	int	node_count =0;
+	int	node_id = 0;
+	int	*nid_arr = NULL;
+	int	atype = READ_WRITE | ATR_DFLAG_CVTSLT;
+	char	mpphost_knl[BASIL_STRING_LONG];
+
+
+	basil_system_element_t *node_group;
+
+
+	if (sys_knl == NULL)
+		return;
+
+	strncpy(mpphost_knl, sys_knl->mpp_host, sizeof(mpphost_knl) - 1);
+	mpphost_knl[sizeof(mpphost_knl) - 1] = '\0';
```

Comment:
Instead of doing a strncpy() here, why don't you do a snprintf and then you won't need to add a NULL at the very end of the buffer.